### ISSUE TITLE
Incorrect "Called C++ object pointer is null" report -- when user code NULL-checks the dynamically casted value

### ISSUE THREAD

**Issue by vakatov** [llvm/llvm-project#134454]
It looks that when the Static Analyzer sees a code branch which checks for a NULL **dynamically casted** value, then it (mistakenly) decides that the original pointer therefore is liable to be NULL as well:

```
class A { public: virtual void f(void) const; };
class B : public A { public: virtual void f(void) const; };

void g(const A* a)
{
    const B* b = dynamic_cast<const B*>(a);
    if (b) {
        b->f();
        return;
    }
    a->f();
}

$ /usr/local/llvm/20.1.1/bin/clang++ --analyze fp.cpp
fp.cpp:11:5: warning: Called C++ object pointer is null [core.CallAndMessage]
   11 |     a->f();
      |     ^~~~~~
1 warning generated.
```

**Comment by vakatov** [comment]
We have quite a few such false positives reported for our code.

N.B. I improved the code in the sample a little (it looked a tad silly before... although it still illustrated the problem).

**Comment by vakatov** [comment]
Great, thanks!

**Comment by vbvictor** [comment]
Closing this issue as it seems to be resolved.
If there are any more questions, please reopen it or submit a new one.

**Comment by vakatov** [comment]
>  Closing this issue as it seems to be resolved.

@vbvictor - do you mean that it's fixed in the contemporary code of the Static Analyzer? (Sorry, I don't see pointers to any related changes, at least looking at this web page.)

**Comment by steakhal** [comment]
This wasn't fixed. The issue with `dyn_casts` is still here. It's just not urgent as there is a fairly practical workaround for it.
At one day one might fix this.

**Comment by vbvictor** [comment]
Huge sorry for the mess, I was going through CSA issues to find a bug to fix. Maybe I had some different issue in mind when I wrote the comment... :(
I'll look into it
